matron/src/device/device_serial.c

```diff
@@ -0,0 +1,188 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/input.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/queue.h>
+#include <termios.h>
+#include <unistd.h>
+#include <lualib.h>
+
+#include "device.h"
+#include "device_serial.h"
+#include "events.h"
+#include "lua_eval.h"
+
+#define ISPEED "ispeed"
+#define OSPEED "ospeed"
+#define IFLAG "iflag"
+#define OFLAG "oflag"
+#define CFLAG "cflag"
+#define LFLAG "lflag"
+#define LINE "line"
+#define CC "cc"
+
+int dev_serial_init(void *self, lua_State *l) {
+    struct dev_serial *d = (struct dev_serial *)self;
+    struct dev_common *base = (struct dev_common *)self;
+
+    d->fd = open(d->base.path, O_RDWR | O_NOCTTY | O_SYNC);
+    if (d->fd < 0) {
+        fprintf(stderr, "dev_serial: failed to open serial device: %s\n", d->base.path);
+        return -1;
+    }
+
+    d->handler_id = strdup(lua_tostring(l, -1));
+    fprintf(stderr, "dev_serial: TTY device %s at %s handled by %s\n", d->base.name, d->base.path, d->handler_id);
+
+    // Initialize a new terminal config using the old config. The new config will
+    // be passed to lua for overrides.
+    tcgetattr(d->fd, &d->oldtio);
+    d->newtio = d->oldtio;
+
+    lua_getglobal(l, "_norns");
+    lua_getfield(l, -1, "serial");
+    lua_remove(l, -2);
+    lua_getfield(l, -1, "configure");
+    lua_remove(l, -2);
+
+    lua_pushstring(l, d->handler_id);
+
+    lua_createtable(l, 0, 8);
+    lua_pushinteger(l, cfgetispeed(&d->oldtio));
+    lua_setfield(l, -2, ISPEED);
+    lua_pushinteger(l, cfgetospeed(&d->oldtio));
+    lua_setfield(l, -2, OSPEED);
+    lua_pushinteger(l, d->oldtio.c_iflag);
+    lua_setfield(l, -2, IFLAG);
+    lua_pushinteger(l, d->oldtio.c_oflag);
+    lua_setfield(l, -2, OFLAG);
+    lua_pushinteger(l, d->oldtio.c_cflag);
+    lua_setfield(l, -2, CFLAG);
+    lua_pushinteger(l, d->oldtio.c_lflag);
+    lua_setfield(l, -2, LFLAG);
+    lua_pushinteger(l, d->oldtio.c_line);
+    lua_setfield(l, -2, LINE);
+
+    lua_createtable(l, NCCS, 0);
+    for (int i = 0; i < NCCS; i++) {
+        lua_pushinteger(l, d->oldtio.c_cc[i]);
+        lua_rawseti(l, -2, i+1);
+    }
```

Comment:
I'm not super familiar with manipulating Lua state, `lua_createtable()` is being prepped for 8 records, but it looks like there might be 9 -- is `NCCS` is part of that table too or does it end with `CC`?

EDIT: Nevermind, appears `NCCS` is a nested table inside `CC`.